src/MassTransit/Transports/ConsumerAgent.cs

```diff
@@ -174,9 +189,22 @@ protected virtual async Task ActiveAndActualAgentsCompleted(StopContext context)
             }
         }
 
-        protected Task Dispatch(ReceiveContext context, ReceiveLockContext receiveLock = default)
+        protected async Task Dispatch<TContext>(TKey key, TContext context, ReceiveLockContext receiveLock = default)
+            where TContext : BaseReceiveContext
         {
-            return _dispatcher.Dispatch(context, receiveLock);
+            var added = _pending.TryAdd(key, context);
+            if (!added)
```

Comment:
Uhm, isn't there a RabbitMQ case where the delivery key is always the same for the direct exchange? Seems like something was dropped here.